### PR TITLE
Delete virtual node when virtual kubelet is removed

### PR DESCRIPTION
**Pull request by sashayakovtseva**
Right now when virtual kubelet pod is removed, virtual node is transitioned in a state `NotReady` and remains in k8s node list. This PR fixes that.

### CODE REVIEW

**Comment by sashayakovtseva** [comment]
Not sure if node should be silently removes so that test passes, or a test should be updated. 
Let me know what it better!

**Comment by cpuguy83** [comment]
Can you point to where the kubelet handles this?

File: .gitignore [hunk 1]
```diff
@@ -33,6 +33,9 @@ loganalytics.json
 # VS Code files
 .vscode/
 
+# IntelliJ Goland files
```
Comment (sargun) [inline .gitignore]: Can you split this into a separate PR?

File: cmd/virtual-kubelet/commands/root/root.go [hunk 2]
```diff
@@ -60,6 +61,9 @@ This allows users to schedule kubernetes workloads on nodes that aren't running
 }
 
 func runRootCommand(ctx context.Context, c Opts) error {
+	var wg sync.WaitGroup
+	defer wg.Wait()
```
Comment (sargun) [inline cmd/virtual-kubelet/commands/root/root.go]: This seems like it'd be very confusing to troubleshoot a deadlock. Why not explicitly move it?

File: cmd/virtual-kubelet/commands/root/root.go [hunk 3]
```diff
@@ -102,8 +106,15 @@ func runRootCommand(ctx context.Context, c Opts) error {
...
 	serviceInformer := scmInformerFactory.Core().V1().Services()
 
-	go podInformerFactory.Start(ctx.Done())
-	go scmInformerFactory.Start(ctx.Done())
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		podInformerFactory.Start(ctx.Done())
+	}()
+	go func() {
```
Comment (sargun) [inline cmd/virtual-kubelet/commands/root/root.go]: Why not use errgroup?

File: vkubelet/node.go [hunk 4]
```diff
@@ -233,7 +233,7 @@ func (n *Node) controlLoop(ctx context.Context) error {
 	for {
 		select {
 		case <-ctx.Done():
-			return nil
+			return n.nodes.Delete(n.n.Name, &metav1.DeleteOptions{})
```
Comment (sargun) [inline vkubelet/node.go]: This seems problematic. What if there are still running pods? Does normal kubelet exhibit this behaviour?

**Comment by sashayakovtseva** [comment]
@cpuguy83 Node monitoring code can be found [here](https://github.com/kubernetes/kubernetes/blob/release-1.13/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L651) if that is what you meant.
As I see k8s doesn't remove node automatically there, also based on experiments node will remain in the list.

File: .gitignore (hunk 1 above)
Comment (sashayakovtseva) [reply .gitignore]: Sure, see #660 please.

File: cmd/virtual-kubelet/commands/root/root.go (hunk 2 above)
Comment (sashayakovtseva) [reply cmd/virtual-kubelet/commands/root/root.go]: For sake of simplicity: wait in one place instead of each return.
Another option would be to use wait group for goroutine with `node.Run` only, but leaving the rest goroutines without attention is not nice, I think.

File: cmd/virtual-kubelet/commands/root/root.go (hunk 3 above)
Comment (sashayakovtseva) [reply cmd/virtual-kubelet/commands/root/root.go]: I haven't heard of it before, can you elaborate? :)
Will it make any difference?

**Comment by cpuguy83** [comment]
I would say then let's follow what the kubelet does.
Since vk is a library, consumers can choose to delete the node or not on shutdown.

**Comment by pires** [comment]
@sashayakovtseva as you correctly pointed out, there's a node lifecycle controller that handles this. That controller, however, is not part of the `kubelet` but of the `controller-manager`. What is the reasoning behind having similar behavior in the VK when the `controller-manager` will do it?

**Comment by sashayakovtseva** [comment]
@pires The initial idea was to make cleanup after pod removal, b/c node was created in pod so in my understanding it should be removed with pod removal as well. 
Does `nodelifecycle` removes `NotReady` nodes? According to source code, it does not, that is why this PR was suggested in the first place.

**Comment by sashayakovtseva** [comment]
@cpuguy83 
> consumers can choose to delete the node or not on shutdown

Do you mean end user will have to remove node manually? Or there is some flag in vk? If not, that is another option how node removal can be implemented.

**Comment by cpuguy83** [comment]
I mean that whatever integration people choose to make, they can handle it outside of the scope of vk core libraries.
I do not think the life-cycle of the node should be tied to a particular pod life-cycle implicitly.
There are good reasons to take down a pod that do not mean the node should go away.
So ultimately, whatever is managing the pod should be responsible for cleaning it up if it shouldn't be there.

**Comment by pires** [comment]
@sashayakovtseva I had missed the fact you're running it as a pod but, still, what's the difference here? Say the `kubelet` didn't exit cleanly, it is marked as `NotReady` and the node stays in the API, right?

I understand what you may want to do here but I still think it shouldn't be the responsibility of the VK to guarantee it exists cleanly. If it doesn't exit cleanly, how can you then guarantee the corresponding node is removed?

**Comment by sashayakovtseva** [comment]
@pires Hm, there are no guarantees in case vk exits abnormally. And the question is, should they be? I mean if it suddenly breaks everything can go wrong, only external monitor can cleanup in that case.

**Comment by sashayakovtseva** [comment]
Thanks everyone for comments and discussion, looks like this PR doesn't bring any value to the virtual-kubelet. Closing this PR, unless anyone else thinks this behavior may be useful.

**Comment by pires** [comment]
@sashayakovtseva but the same applies to the kubelet, right? Maybe I'm missing something, and if that's the case, please, do point it out. Always learning and I've taken lessons from this thread already, so thank you for your contribution(s).

**Comment by sashayakovtseva** [comment]
@pires Yes, when kubelet falls down, node is left in the list afaik. Maybe you are right and there is no much point in cleaning up node from vk pod. 
However, we will use node removal in our fork as it looks convenient for us :) 
Thank you for pointing things out!
